### DISCUSSION TITLE
Loading fonts from file

### DISCUSSION THREAD

**Discussion by alo00** [AvaloniaUI/Avalonia#13563] [Q&A]
Hello, I have been able to use custom fonts by including it in the resources of the project.

Is there any way to load fonts directly from file at runtime?

**Comment by thevortexcloud** [comment]
I am not sure if there is a better way (I have never touched custom fonts), but have you tried the runtime XAML loader? It lets you load and parse XAML files at runtime. Which would in theory let you add them as resources using code behind.

This is probably still not as direct as you would like though. Since it requires you to define some sort of resource XAML file.

**Comment by timunie** [comment]
I guess the `FontManager` should be able to load fonts at runtime.

**Reply by Gillibald** [reply]
This is possible via custom FontCollection

**Reply by Gillibald** [reply]
Sample implementation: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/Fonts/EmbeddedFontCollection.cs

**Reply by alo00** [reply]
Tanks a lot, I will have a look into this.

**Reply by alo00** [reply]
Hello.  I think your pointing me to the wrong sample.   EmbeddedFontCollection works fine for loading fonts from resource, not from a ttf file in a folder.  Unless there is a way to have the IAssetLoader loading from a file. Is it possible ?

**Reply by Gillibald** [reply]
The sample shows how to load fonts via stream and make them available via a custom font collection

**Reply by taooceros** [reply]
@Gillibald some method used in EmbeddedFontCollection are internal to Avalonia. I wonder how to get around with that when implement a custom font collection?

**Reply by Gillibald** [reply]
You are supposed to implement this by subclassing https://github.com/AvaloniaUI/Avalonia/blob/c3911177956897d7a5176f9d59932b308aebc0c9/src/Avalonia.Base/Media/Fonts/FontCollectionBase.cs#L11

**Reply by Gillibald** [reply]
The EmbeddedFontCollection is some example. Feel free to copy it and instead of using the asset loader load the file via File API

**Reply by taooceros** [reply]
The issue is the api used in `EmbeddedFontCollection` is `internal` to Avalonia, which means the subtype I create cannot access them. See #14280 for which are the api that is important

**Reply by Gillibald** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/c3911177956897d7a5176f9d59932b308aebc0c9/src/Avalonia.Base/Platform/IFontManagerImpl.cs#L61 isn't internal so you got all you need. What am I missing?

**Reply by taooceros** [reply]
huh very weird that Rider indicate that it is internal...
![image](https://github.com/AvaloniaUI/Avalonia/assets/45326534/15f4d444-6d26-489a-84a8-91cad075410d)

**Reply by appel1** [reply]
> https://github.com/AvaloniaUI/Avalonia/blob/c3911177956897d7a5176f9d59932b308aebc0c9/src/Avalonia.Base/Platform/IFontManagerImpl.cs#L61
> 
> isn't internal so you got all you need. What am I missing?

It is internal in the ref assembly which is used at compile time. 

![image](https://github.com/AvaloniaUI/Avalonia/assets/973314/d637442e-e406-4782-aba3-b33e0e40b9e5)

![image](https://github.com/AvaloniaUI/Avalonia/assets/973314/d6454c0f-ad19-4d10-aed4-1b092956169d)

To get font loading to work when building a browser app I've had to resort to reflection for now.

**Reply by appel1** [reply]
You can enable access by setting AvaloniaAccessUnstablePrivateApis to true in the project file instead of reflection. 

https://github.com/AvaloniaUI/Avalonia/issues/18777

**Comment by xinghaiyege** [comment]
```c#
        var font_family = "Font Name";
        FontFamilyLoader.LoadFontAssets(new Uri("font.otf"));
        Setter setter = new()
        {
            Property = FontFamilyProperty,
            Value = new FontFamily(font_family)
        };
        Style windows_style = new(x => x.OfType<Window>());
        windows_style.Add(setter);
        Application.Current.Styles.Add(windows_style);
        Style context_menu_style = new(x => x.OfType<ContextMenu>());
        context_menu_style.Add(setter);
        Application.Current.Styles.Add(context_menu_style);
        Style tooltip_style = new(x => x.OfType<ToolTip>());
        tooltip_style.Add(setter);
        Application.Current.Styles.Add(tooltip_style);
```

**Comment by Gillibald** [comment]
I will introduce support for file and http scheme for custom fonts

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/issues/14280

**Comment by alanthinker** [comment]
Can modify like this to support local file:

EmbeddedFontCollection class:
```
public override void Initialize(IFontManagerImpl fontManager)
{
    if (_source.IsFile)
    {
        var stream = File.Open(_source.LocalPath, FileMode.Open);

        if (fontManager.TryCreateGlyphTypeface(stream, out var glyphTypeface))
        {
            if (!_glyphTypefaceCache.TryGetValue(glyphTypeface.FamilyName, out var glyphTypefaces))
            {
                glyphTypefaces = new ConcurrentDictionary<FontCollectionKey, IGlyphTypeface?>();

                if (_glyphTypefaceCache.TryAdd(glyphTypeface.FamilyName, glyphTypefaces))
                {
                    _fontFamilies.Add(new FontFamily(_key, glyphTypeface.FamilyName));
                }
            }

            var key = new FontCollectionKey(
                   glyphTypeface.Style,
                   glyphTypeface.Weight,
                   glyphTypeface.Stretch);

            glyphTypefaces.TryAdd(key, glyphTypeface);
        }
    }
    else
    {
        var assetLoader = AvaloniaLocator.Current.GetRequiredService<IAssetLoader>();

        var fontAssets = FontFamilyLoader.LoadFontAssets(_source);

        foreach (var fontAsset in fontAssets)
        {
            var stream = assetLoader.Open(fontAsset);

            if (fontManager.TryCreateGlyphTypeface(stream, out var glyphTypeface))
            {
                if (!_glyphTypefaceCache.TryGetValue(glyphTypeface.FamilyName, out var glyphTypefaces))
                {
                    glyphTypefaces = new ConcurrentDictionary<FontCollectionKey, IGlyphTypeface?>();

                    if (_glyphTypefaceCache.TryAdd(glyphTypeface.FamilyName, glyphTypefaces))
                    {
                        _fontFamilies.Add(new FontFamily(_key, glyphTypeface.FamilyName));
                    }
                }

                var key = new FontCollectionKey(
                       glyphTypeface.Style,
                       glyphTypeface.Weight,
                       glyphTypeface.Stretch);

                glyphTypefaces.TryAdd(key, glyphTypeface);
            }
        }
    }
}
```

FontManager class
```
        private bool TryGetFontCollection(Uri source, [NotNullWhen(true)] out IFontCollection? fontCollection)
        {
            if (source.Scheme == SystemFontScheme)
            {
                source = SystemFontsKey;
            }

            if (!_fontCollections.TryGetValue(source, out fontCollection) && (source.IsAbsoluteResm() || source.IsAvares() || source.IsFile))
            {
                var embeddedFonts = new EmbeddedFontCollection(source, source);

                embeddedFonts.Initialize(PlatformImpl);

                if (embeddedFonts.Count > 0 && _fontCollections.TryAdd(source, embeddedFonts))
                {
                    fontCollection = embeddedFonts;
                }
            }

            return fontCollection != null;
        }
```

**Reply by alanthinker** [reply]
Then you can use:
1.
```
  <Application.Resources>
    <FontFamily x:Key="myfont">file:///C:\Users\Alan\source\repos\AvaloniaApplication11\AvaloniaApplication11\Assets\abc.ttf#CN-Hanh Khai</FontFamily>
  </Application.Resources>
```

2.
```
var font = new FontFamily(null, @"file:///C:\Users\Alan\source\repos\AvaloniaApplication11\AvaloniaApplication11\Assets\abc.ttf#CN-Hanh Khai");
```

3.
```
var filePth = @"file:///C:\Users\Alan\source\repos\AvaloniaApplication11\AvaloniaApplication11\Assets\abc.ttf";
var fileUri = new Uri(filePth);
var xx = new EmbeddedFontCollection(fileUri, fileUri);
xx.Initialize(FontManager.Current.PlatformImpl);
var fontName = xx.FirstOrDefault().FamilyNames;
var fullFontPath = filePth + "#" + fontName;
var font = new FontFamily(null, fullFontPath);
```
